onnxruntime/core/providers/vitisai/imp/ep_context_utils.cc

```diff
@@ -0,0 +1,831 @@
+  p_attr_0->set_type(ONNX_NAMESPACE::AttributeProto::INT);
+  p_attr_0->set_i(embed_mode);
+  // Attr "ep_cache_context".
+  auto p_attr_1 = ONNX_NAMESPACE::AttributeProto::Create();
+  p_attr_1->set_name(kEPCacheContextAttr);
+  // p_attr_1->set_type(onnx::AttributeProto_AttributeType_STRING);
+  p_attr_1->set_type(ONNX_NAMESPACE::AttributeProto::STRING);
+  // Relative to the ONNX model file.
+  p_attr_1->set_s(
+      embed_mode == 0 ? fs::path(ctx_cache_file_loc).filename().string() : serialized_ctx_cache);
+  // Attr "source".
+  auto p_attr_2 = ONNX_NAMESPACE::AttributeProto::Create();
+  p_attr_2->set_name(kSourceAttr);
+  // p_attr_2->set_type(onnx::AttributeProto_AttributeType_STRING);
+  p_attr_2->set_type(ONNX_NAMESPACE::AttributeProto::STRING);
+  p_attr_2->set_s(kVitisAIExecutionProvider);
+  // Attr "onnx_model_filename".
+  auto p_attr_3 = ONNX_NAMESPACE::AttributeProto::Create();
+  p_attr_3->set_name(kONNXModelFileNameAttr);
+  // p_attr_3->set_type(onnx::AttributeProto_AttributeType_STRING);
+  p_attr_3->set_type(ONNX_NAMESPACE::AttributeProto::STRING);
+  p_attr_3->set_s(fs::path(graph_viewer.ModelPath().ToPathString()).filename().string());
+  // Attr "notes".
+  auto p_attr_4 = ONNX_NAMESPACE::AttributeProto::Create();
+  p_attr_4->set_name(kNotesAttr);
+  // p_attr_4->set_type(onnx::AttributeProto_AttributeType_STRING);
+  p_attr_4->set_type(ONNX_NAMESPACE::AttributeProto::STRING);
+  // FIXME: 2G-limit of ProtoBuf.
+  if (saving_orig_graph) {
+    p_attr_4->set_s(SerializeOrigialGraph(graph_viewer));
+    LOGS_DEFAULT(VERBOSE) << "Saved serialized graph to attr proto";
+  } else {
+    nlohmann::json j_obj;
+    j_obj["backend_cache_dir"] = backend_cache_dir;
+    j_obj["backend_cache_key"] = backend_cache_key;
+    p_attr_4->set_s(j_obj.dump(-1, ' ', false, nlohmann::json::error_handler_t::replace));
+    LOGS_DEFAULT(VERBOSE) << "Saved backend cache key to attr proto";
+  }
+
+  auto p_node_attrs = NodeAttributes::Create();
+  constexpr int num_attrs = 5;
+  p_node_attrs->reserve(num_attrs);
+  p_node_attrs->emplace(kEmbedModeAttr, *p_attr_0);
+  p_node_attrs->emplace(kEPCacheContextAttr, *p_attr_1);
+  p_node_attrs->emplace(kSourceAttr, *p_attr_2);
+  p_node_attrs->emplace(kONNXModelFileNameAttr, *p_attr_3);
+  p_node_attrs->emplace(kNotesAttr, *p_attr_4);
+
+  // Since we don't implement `IExecutionProvider::GetEpContextNodes()` and
+  // thus don't leverage `CreateEpContextModel()` in the file "graph_partitioner.cc",
+  // we specify a brand-new node name here.
+  ep_ctx_graph.AddNode(kEPContextOpName, kEPContextOp, "", input_node_arg_ptrs, output_node_arg_ptrs, p_node_attrs.get(), kEPContextOpDomain);
+
+  auto res_status = ep_ctx_graph.Resolve();
+  ORT_ENFORCE(res_status.IsOK(), res_status.ErrorMessage());
+  LOGS_DEFAULT(VERBOSE) << "Created EP context model graph resolved";
+
+  auto p_ep_ctx_graph_viewer = ep_ctx_graph.CreateGraphViewer();
+  auto p_temp_model_2 = p_ep_ctx_graph_viewer->CreateModel(*p_logger);
```

Comment:
why need to create the model again?

Reply:
1) Based on the documentation "" I posted above, creating another model again looks redundent.

2) I was following the way TRT EP is doing at https://github.com/microsoft/onnxruntime/blob/4743803944d6f6333bf93b89a62d8083d4466710/onnxruntime/core/providers/tensorrt/onnx_ctx_model_helper.cc#L130-L133

3) It's supposed the way below should be working as well - without creating one more model `p_temp_model_2`:
```
auto p_ep_ctx_graph_viewer = ep_ctx_graph.CreateGraphViewer();
auto p_ep_ctx_model_proto = p_temp_model->ToProto();
p_ep_ctx_graph_viewer->ToProto(*p_ep_ctx_model_proto->mutable_graph(), true, true);
```
However, IIRC, I tried this way out but encountered some issues when tesing.